### PR TITLE
CI: Setup Qodana workflow

### PR DESCRIPTION
**Pull request by mchernyavsky**
changelog: Setup [Qodana](https://www.jetbrains.com/qodana/) checks on CI

Screen Shot 2022-09-20 at 13 33 25

Screen Shot 2022-09-20 at 13 33 36

Screen Shot 2022-09-20 at 15 52 07

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt [hunk 1]
```diff
@@ -1235,17 +1235,6 @@ class ImplLookup(
...
-    fun isAnyFn(ty: Ty): Boolean = isFn(ty) || isFnOnce(ty) || isFnMut(ty)
-
-    @Suppress("MemberVisibilityCanBePrivate")
-    fun isFn(ty: Ty): Boolean = ty.isTraitImplemented(items.Fn)
-
-    @Suppress("MemberVisibilityCanBePrivate")
-    fun isFnOnce(ty: Ty): Boolean = ty.isTraitImplemented(items.FnOnce)
-
-    @Suppress("MemberVisibilityCanBePrivate")
-    fun isFnMut(ty: Ty): Boolean = ty.isTraitImplemented(items.FnMut)
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt:1247]: This API may be useful
@vlad20012 what do you think?

File: .github/workflows/code_quality.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
+name: code quality
+on:
+    push:
+        branches:
+            - master
+            - staging
+            - trying
```
Comment (Undin) [inline .github/workflows/code_quality.yml:7]: Do we really need these checks for these branches? I can't come up with when it can be useful

**Review by Undin — COMMENTED** [review]
Could you somehow show how it works when there are warnings?
Also, it would be nice to add changelog section to PR description

File: src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt (hunk 1 above)
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/ImplLookup.kt:1247]: No, Let's remove

File: src/main/kotlin/org/rust/lang/core/macros/MacroExpansionFileSystem.kt [hunk 3]
```diff
@@ -500,7 +500,6 @@ class MacroExpansionFileSystem : NewVirtualFileSystem() {
     }
 
     fun exists(path: String): Boolean = convert(path) != null
-    fun isFile(path: String): Boolean = convert(path) is FSFile
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/macros/MacroExpansionFileSystem.kt:503]: Seems like we forget to remove it after rewriting the macro expansion system, so no problems to remove it now

File: qodana.yaml [hunk 4]
```diff
@@ -1,9 +1,10 @@
-version: 1.0
+version: "1.0"
+linter: jetbrains/qodana-jvm-community:2022.2
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply qodana.yaml]: I have removed this line, but we have to update Qodana GitHub Action version in the workflow file from time to time.

File: .github/workflows/code_quality.yml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+            - name: Download
+              uses: gradle/gradle-build-action@v1
+              with:
+                  arguments: ":resolveDependencies -Pkotlin.incremental=false --no-daemon"
+
+            - name: Build
+              uses: gradle/gradle-build-action@v1
+              with:
+                  arguments: "assemble testClasses -Pkotlin.incremental=false --no-daemon"
```
Comment (Undin) [inline .github/workflows/code_quality.yml:26]: Do we really need to compile code? I thought the main idea of Qodana is to analyze the code without compilation

**Comment by mchernyavsky** [comment]
@Undin This PR is ready for merge :)

**Review by Undin — CHANGES_REQUESTED** [review]
I see the following warning in my IDE for `RsGoToDeclarationRunningService` 
image

It seems the Qodana doesn't report it at all. I don't think it should stop us from enabling Qodana but it requires some investigation where is the problem: Qodana itself or inspection profile

File: src/main/kotlin/org/rust/lang/core/stubs/index/RsReexportIndex.kt [hunk 6]
```diff
@@ -46,27 +42,6 @@ class RsReexportIndex : AbstractStubIndex<ReexportKey, RsUseSpeck>() {
             originalName?.let { sink.occurrence(KEY, ReexportKey.OriginalNameKey(it)) }
             producedName?.let { sink.occurrence(KEY, ReexportKey.ProducedNameKey(it)) }
         }
-
-        fun findReexportsByProducedName(
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply src/main/kotlin/org/rust/lang/core/stubs/index/RsReexportIndex.kt:50]: Isn't it used in `IndexSink.indexUseSpeck`, which is used in `RsUseSpeckStub`?

File: qodana.yaml [hunk 7]
```diff
@@ -1,9 +1,12 @@
...
+  name: qodana.recommended
 exclude:
-    - name: UnstableApiUsage
-    - name: UnstableTypeUsedInSignature
-    - name: IncompleteDestructuring
-    - name: KDocUnresolvedReference
-    - name: DialogTitleCapitalization
+  - name: DialogTitleCapitalization
+  - name: IncompleteDestructuring
+  - name: KDocUnresolvedReference
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply qodana.yaml:8]: I will commit the version without this line so you can check the warnings that Qodana produces.

File: src/main/kotlin/org/rust/lang/core/stubs/index/RsReexportIndex.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/stubs/index/RsReexportIndex.kt:50]: It's true that we calculate some index for `use` items
But we don't use its results anywhere anymore. Previously, we used the index for auto-import. But now auto-import feature uses info collected by name resolution

File: qodana.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply qodana.yaml]: We have 10 "false-positives" due to the usage of `&[array]`s and `#[attr]`s in docs. I can suppress them individually. What do you think?

[the remark this answers is not in the document]
Comment (Undin) [reply qodana.yaml]: 10 isn't a lot, especially taking into account that 5 of them in the single comment
Let's suppress them

File: .github/workflows/code-quality.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+            - name: Generate sources
+              uses: gradle/gradle-build-action@v1
+              with:
+                  arguments: ":generateLexer :generateParser debugger:generateGrammarSource debugger:generateTestGrammarSource -Pkotlin.incremental=false --no-daemon"
+
+            - name: Qodana
+              uses: JetBrains/qodana-action@v2022.3.0
+              with:
+                  use-caches: false
```
[the remark this answers is not in the document]
Comment (mchernyavsky) [reply .github/workflows/code-quality.yml:40]: Now it analyzes only changed files.

**Review by Undin — APPROVED** [review]
🎉 

bors r+

**Pull request merged**